parsing/location.ml

```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+          end;
+          if is_start pos then begin
+            highlighted := true;
+            Format.fprintf ppf "@{<%s>" highlight_tag;
+          end;
+          if !highlighted then Format.pp_print_char ppf '^'
+          else Format.pp_print_char ppf ' '
+        done;
+        Format.fprintf ppf "@}@,"
+    | _ ->
+        (* Multi-line error *)
+        let ellipsed_first, ellipsed_last =
+          let lines_nb = List.length lines in
+          if lines_nb > max_lines then
+            let printed_lines = max_lines - 1 in (* ellipsis uses one line *)
+            let lines_before = printed_lines / 2 + printed_lines mod 2 in
+            let lines_after = printed_lines / 2 in
+            (lines_before, lines_nb - lines_after - 1)
+          else (-1, -1)
+        in
+        let highlighted = ref false in
+        List.iteri (fun k (line, line_nb_opt, line_start_cnum) ->
+          if k = ellipsed_first then
+            Format.fprintf ppf "%s@," (String.make (lines_nb_maxlen + 2) '.');
+          if ellipsed_first <= k && k <= ellipsed_last then ()
+          else begin
+            Format.fprintf ppf "%a" (prefix ~sep:true) line_nb_opt;
+            for i = 0 to String.length line - 1 do
+              if is_end (line_start_cnum + i) then highlighted := false;
+              if is_start (line_start_cnum + i) then highlighted := true;
+              if !highlighted then Format.pp_print_char ppf line.[i]
+              else Format.pp_print_char ppf '.'
+            done;
+            Format.fprintf ppf "@,"
+          end
+        ) lines;
     end;
-    Format.fprintf ppf "@]@,@]"
+    Format.fprintf ppf "@]"
   end
 
-let highlight_dumb lb ppf locs =
-  try highlight_dumb lb ppf locs with Exit -> ()
+let lines_around
+    ~(start_pos: position) ~(end_pos: position)
+    ~(seek: int -> unit)
+    ~(read_char: unit -> char option):
+  (string (* line *) * int (* beginning of line *)) list
+  =
+  seek start_pos.pos_bol;
+  let lines = ref [] in
+  let bol = ref start_pos.pos_bol in
+  let cur = ref start_pos.pos_bol in
+  let b = Buffer.create 80 in
+  let add_line () =
+    if !bol < !cur then begin
+      let line = Buffer.contents b in
+      Buffer.clear b;
+      lines := (line, !bol) :: !lines;
+      bol := !cur
+    end
+  in
+  let rec loop () =
+    if !bol >= end_pos.pos_cnum then ()
+    else begin
+      match read_char () with
+      | None ->
+          (* end of input *)
+          add_line ()
+      | Some c ->
+          incr cur;
+          match c with
+          | '\r' -> loop ()
+          | '\n' -> add_line (); loop ()
+          | _ -> Buffer.add_char b c; loop ()
+    end
+  in
+  loop ();
+  List.rev !lines
+
+(* Try to get lines from a lexbuf *)
+let lexbuf_lines_around
+    ~(start_pos: position) ~(end_pos: position)
+    (lb: lexbuf):
+  (string * int) list
+  =
+  (* Converts a global position to one that is relative to the lexing buffer *)
+  let rel n = n - lb.lex_abs_pos in
```

Comment:
Hum, I suspect this is in general incorrect, because I'm not taking `lb.lex_start_pos` into account?

Reply:
(cc @diml or @let-def)

Reply:
If I read the code of lexing.ml correctly, I believe it should indeed be something like: `n - lb.lex_abs_pos + lb.lex_start_pos`. Does the current code work?

BTW, I find looking directly into the lexing buffer internals a bit tedious. Personally, I would just keep the whole input as a single string and extract the source code from that string, that would make the code simpler.

Reply:
The code that processes an input file. But yh, I agree that for plain files it's not an issue. For the toplevel, the code would need to be adapted. Without changing everything, we could keep all the input read since the start of the current loop iteration in a `Buffer.t`.

Reply:
@diml I just tried your suggestion, and it breaks the function (either stops quoting the source, or quotes wrong parts of it) :D. So I guess the current implementation is correct?

Reply:
I guess so. Thanks for trying, at least we know the current one is more likely to be right :)